<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:unless value="$(arg sim_mode)">
        <ros2_control name="physical_robot" type="system">
            <hardware>
            <plugin>robot_pkg/DiffBotSystemHardware</plugin>
                <param name="left_wheel_name">left_wheel_joint</param>
                <param name="right_wheel_name">right_wheel_joint</param>
                <param name="motor_step_per_revolution">1600</param>
                <param name="encoder_step_per_rev">4096</param>
                <param name="invert_motor_left">false</param>
                <param name="invert_motor_right">true</param>
                <param name="invert_encoder_left">false</param>
                <param name="invert_encoder_right">true</param>
                <param name="swap_motors">false</param>
                <param name="swap_encoders">false</param>
                <param name="motors_acceleration">1600</param>
                <param name="encoders_read_frequency">20.0</param>
                <param name="encoder_wheel_diameter">0.0472</param>
                <param name="encoder_wheel_separation">0.1430</param>
                <param name="motor_wheel_diameter">0.0472</param>
                <param name="motor_wheel_separation">0.131</param>
            </hardware>
            <joint name="left_wheel_joint">
                <command_interface name="velocity">
                    <max>3</max>
                    <min>-3</min>               
                </command_interface>
                <state_interface name="position"/>
                <state_interface name="velocity"/>
            </joint>
            <joint name="right_wheel_joint">
                <command_interface name="velocity">
                    <max>3</max>
                    <min>-3</min>             
                </command_interface>
                <state_interface name="position"/>
                <state_interface name="velocity"/>
            </joint>
        </ros2_control>
    </xacro:unless>

    <xacro:if value="$(arg sim_mode)">
        <ros2_control name="GazeboSystem" type="system">
            <hardware>
                <plugin>gazebo_ros2_control/GazeboSystem</plugin>
            </hardware>
            <joint name="left_wheel_joint">
                <command_interface name="velocity">
                    <param name="min">-3</param>
                    <param name="max">3</param>
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/>
            </joint>
            <joint name="right_wheel_joint">
                <command_interface name="velocity">
                    <param name="min">-3</param>
                    <param name="max">3</param>
                </command_interface>
                <state_interface name="velocity"/>
                <state_interface name="position"/>
            </joint>

            <!-- <joint name="encoder_left_wheel_joint">
                <state_interface name="velocity"/>
                <state_interface name="position"/>
            </joint>
            <joint name="encoder_right_wheel_joint">
                <state_interface name="velocity"/>
                <state_interface name="position"/>
            </joint> -->
        </ros2_control>
    </xacro:if>

    <gazebo>
        <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
            <parameters>$(find robot_pkg)/config/diff_drive_controller.yaml</parameters>
            <parameters>$(find robot_pkg)/config/gaz_ros2_ctl_use_sim.yaml</parameters>
        </plugin>
    </gazebo>

</robot>
